serif_shared/src/commonMain/kotlin/xyz/room409/serif/serif_shared/MatrixClient.kt

```diff
@@ -2,82 +2,41 @@ package xyz.room409.serif.serif_shared
-}
-@Serializable
-data class EventIdResponse(val event_id: String)
+class MatrixSession(val client: HttpClient, val access_token: String) {
+    var sync_response: SyncResponse? = null
 
+    // rooms is a pair of room_id and room display name
+    // Display name is calculated by first checking for a room name event in the state events,
+    // then for a canonical alias in the state events, then a list of hero users in the room summary,
+    // and then the error message. Technically we should also be looking at timeline events
+    // for room name changes too. Also, it's not working when there's not a room name -
+    // the way I'm reading the doc heros should be non-null... maybe it's not decoding right?
+    val rooms: List<Pair<String,String>>
+        get() = sync_response?.rooms?.join?.entries?.map { (id, room) ->
+            Pair(id, (room.state.events.firstOfType<StateEvent<RoomNameContent>>()?.content?.name
+                   ?: room.state.events.firstOfType<StateEvent<RoomCanonicalAliasContent>>()?.content?.alias
+                   ?: room.summary.heroes?.joinToString(", ")
+                   ?: "<no room name or heroes>"))
+        }?.toList() ?: listOf()
 
-sealed class MatrixState {
-    val version: String
-        get() {
-            return "Serif Matrix client, pre-alpha on ${Platform().platform}"
-        }
-}
-class MatrixLogin(val login_message: String, val mclient: MatrixClient): MatrixState() {
-    constructor(): this(login_message="Please enter your username and password\n",
-                        mclient=MatrixClient())
-    fun login(username: String, password: String): MatrixState {
-        when (val loginResult = mclient.login(username, password)) {
-            is Success -> { return MatrixRooms(msession=loginResult.value) }
-            is Error -> { return MatrixLogin(login_message="${loginResult.message} - exception was ${loginResult.cause}, please login again...\n",
-                                             mclient=mclient) }
-        }
-    }
-}
-class MatrixRooms(val msession: MatrixSession): MatrixState() {
-    fun getRoom(): MatrixState {
-        return MatrixChatRoom(msession)
-    }
-}
-class MatrixChatRoom(val msession: MatrixSession): MatrixState() {
-    fun sendMessage(msg : String): MatrixState {
-        when (val sendMessageResult = msession.sendMessage(msg)) {
-            is Success -> { println("${sendMessageResult.value}") }
-            is Error -> { println("${sendMessageResult.message} - exception was ${sendMessageResult.cause}") }
-        }
-        return this
-    }
-    fun exitRoom(): MatrixState {
-        msession.closeSession()
-        return MatrixLogin("Closing session, returning to the login prompt for now\n", MatrixClient())
-    }
-}
-class MatrixSession(val client: HttpClient, val access_token: String) {
-    fun sendMessage(msg : String): Outcome<String> {
+    fun sendMessage(msg: String, room_id: String): Outcome<String> {
```

Comment:
I'm loving the lack of hard coded room_ids now.

Reply:
Heck yeah! Getting somewhere!